### ISSUE TITLE
Empty BE Sitemap-List (Ajax XMLHttpRequest Error: Status 500)

### ISSUE THREAD

**Issue by the-robman** [mblaschke/TYPO3-metaseo#225]
Hi @mblaschke . First of all, big thx for this great ext!!! I found a bug (?) in the new metaseo TER release 2.0.0. The BE-Sitemap overview show, that 300+ url's are in database. After updating from the version 1.0.8, the BE-module SEO/sitemap don't show any url's. If I try to reload the list of url's with the reload button, comes a http status 500 over the ajax.php and the list remains empty.

#### System environment:
TYPO3 6.2.19, PHP 5.5.28, index.php is not a symlink, no .htaccess error

#### Ext. implementation:
metaseo 2.0 installed, use update script, empty db caches, empty typo3temp folder, compare database

#### Ajax.php request:
```
Host: www.xxx.de
User-Agent: Mozilla/5.0 (Windows NT 6.1; WOW64; rv:45.0) Gecko/20100101 Firefox/45.0
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8
Accept-Language: de,en-US;q=0.7,en;q=0.3
Accept-Encoding: gzip, deflate
X-Tx-Metaseo-Ajax: 1
X-Requested-With: XMLHttpRequest
Content-Type: application/x-www-form-urlencoded; charset=UTF-8
Referer: http://www.xxx.de/typo3/mod.php?M=MetaseoMetaseo_MetaseoSitemap&moduleToken=082d806bd4a02ceb60c0ba6b22b174b71840ab44&tx_metaseo_metaseometaseo_metaseositemap%5BpageId%5D=1&tx_metaseo_metaseometaseo_metaseositemap%5Baction%5D=sitemap&tx_metaseo_metaseometaseo_metaseositemap%5Bcontroller%5D=BackendSitemap
Content-Length: 191
```

Maybe have you any idea why this 500 error code happens? Greetings, Rob

Edit 1:
On more 500 error in BE menu: Web -> SEO/Meta Tags -> Metadata

Edit 2:
I found an similar issue here https://github.com/mblaschke/TYPO3-metaseo/issues/129
In my case is this JS file loaded: /typo3/contrib/extjs/adapter/ext/ext-base-debug.js wich include the comment "This file is part of Ext JS 3.4".

**Comment by thomaszbz** [comment]
Just tried to reproduce this against TYPO3 6.2.19, but I  can load the entries in SEO/Sitemap. I also can load  Web -> SEO/Meta Tags -> Metadata. For the moment, I cannot (yet) reproduce this bug report.

Related:

- #210 if mixed content is an issue here (should be fixed already in 2.0.0)
- #129 http status 500 for one of TYPO3's breaking changes in respect to Ajax (should be fixed in 2.0.0)
- #93 http status 404 for hidden pages (not fixed yet, also not status 500 as reported)


Please try:

- Clear all cache in the install tool
- Clear Opcode cache in the install tool
- Clear system caches (needs to be enabled in the install tool)
- do that twice again (I can tell you this sometimes helps)

Does the issue still show up?

Can you send us a representation of the http-status 500 _responses_, including headers and content, please (press F12 in FF or Chrome)? Maybe you can see more in debug mode, maybe an exception or an error? Please check all the logs.

**Comment by the-robman** [comment]
I clear "all" caches many many times (fe, general, system, no opcode installed - so no cache files in there, install tool all cache, browser cache), delete the complete typo3temp folder and create a new folder structure with the install tool, no proxy an only one backend open. Still the same problem with the ajax call.
#### Chrome console log
```
POST Request URL:
http://www.xxx.de/typo3/ajax.php?ajaxID=tx_metaseo_controller_ajax_sitemap%3A%3Aindex&ajaxToken=1914aa436fb0ddd9538c26d59fb7a4778a8bb551 500 (Internal Server Error)
asyncRequest @ ext-base-debug.js:2429
Ext.lib.Ajax.pub.request @ ext-base-debug.js:2484
Ext.extend.request @ ext-all-debug.js:4636
Ext.extend.doRequest @ ext-all-debug.js:25841
Ext.extend.request @ ext-all-debug.js:25475
Ext.data.Store.Ext.extend.execute @ ext-all-debug.js:24411
Ext.data.Store.Ext.extend.load @ ext-all-debug.js:24310
Ext.PagingToolbar.Ext.extend.doLoad @ ext-all-debug.js:32842
Ext.PagingToolbar.Ext.extend.doRefresh @ ext-all-debug.js:32871
Ext.Button.Ext.extend.onClick @ ext-all-debug.js:31872h @ ext-all-debug.js:5215
```
ext-base-debug Path: /typo3/contrib/extjs/adapter/ext/ext-base-debug.js
ext-all-debug Path: /typo3/contrib/extjs/ext-all-debug.js
#### Response Headers
```
Accept-Ranges:bytes
Connection:close
Content-Length:1680
Content-Type:text/html
Date:Sat, 12 Mar 2016 17:51:10 GMT
ETag:"690-506679282d840"
Last-Modified:Mon, 27 Oct 2014 13:33:13 GMT
Server:Apache
```
#### Request Headers
```
Accept:*/*
Accept-Encoding:gzip, deflate
Accept-Language:de-DE,de;q=0.8,en-US;q=0.6,en;q=0.4
Cache-Control:no-cache
Connection:keep-alive
Content-Length:191
Content-Type:application/x-www-form-urlencoded; charset=UTF-8
Cookie:PHPSESSID=12435076751f40544d52c83793464f03; fe_typo_user=91b07a8604fe052dc66a4012b517c351; _pk_id.1.53bc=0ac1dd5387e1a4c6.1453843348.2.1457804183.1457804183.; _pk_ses.1.53bc=*; be_typo_user=bb33f16e0eb1621b3d6b565f7438696f; Typo3InstallTool=40d0786165223472a458e0ddb29e99e1
Host:www.xxx.de
Origin:http://www.xxx.de
Pragma:no-cache
Referer:http://www.xxx.de/typo3/mod.php?M=MetaseoMetaseo_MetaseoSitemap&moduleToken=292bc38016353ef2bc3154915e6bfa7b44a72292&tx_metaseo_metaseometaseo_metaseositemap%5BpageId%5D=1&tx_metaseo_metaseometaseo_metaseositemap%5Baction%5D=sitemap&tx_metaseo_metaseometaseo_metaseositemap%5Bcontroller%5D=BackendSitemap
User-Agent:Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/49.0.2623.87 Safari/537.36
X-Requested-With:XMLHttpRequest
X-Tx-Metaseo-Ajax:1
```
#### Query String Parameters
```
ajaxID:tx_metaseo_controller_ajax_sitemap::index
ajaxToken:1914aa436fb0ddd9538c26d59fb7a4778a8bb551
```
#### Form Data
```
start:0
limit:50
sort:uid
dir:DESC
pid:1
pagerStart:0
pagingSize:50
criteriaFulltext:""
criteriaPageUid:""
criteriaPageLanguage:""
criteriaPageDepth:""
criteriaIsBlacklisted:0
```
The funny thing is, that no error occures on my local xampp, with the same environment (apache/mysql/php/typo3/metaseo).

**Comment by thomaszbz** [comment]
Concerning the system which throws http status 500 errors:

Which System Environment do you have?

- Web server and version?
- OS (Distro and version)?

Which non-core TYPO3 extensions do you have installed on that system? Does the list differ from your XAMPP instance?

Important: I still have not seen the _content_ of the _response_ with http status 500. This response could contain an error message or an error code. You might want to do that in TYPO3's debug mode (to be enabled in the install tool temporarily).

**Comment by the-robman** [comment]
Sorry, I cant find any version info for server and os over the phpinfo().
- Online-System: Linux, Apache (SERVER_SOFTWARE), PHP/5.5.28
- Offline-System: Windows, Apache/2.4.3 (Win32), PHP/5.5.28

Online are the same TER Extensions like on my local xampp.
Thats the 500 status message from jweiland hosting:
```
Error 500 - Interner Serverfehler

Beim Aufruf der Seite ist ein Fehler aufgetreten: Interner Serverfehler.

Die aufgerufene Seite oder das Skript haben einen Fehler verursacht.

Hinweis für den Webmaster:

Die häufigsten Ursachen für diese Fehlermeldung sind:

1. die index.php Datei im Startverzeichnis ist ein Symlink auf die tatsächliche index.php Datei. Löschen Sie den Symlink index.php im Startverzeichnis und ersetzen diese durch eine Kopie der index.php Datei im Startverzeichnis des TYPO3 Quellcodes (z.B. typo3_src-4.1.1/index.php)

2. die Datei .htaccess enthält einen Syntaxfehler. Bitte überprüfen Sie die Schreibweise der Befehle (z.B. Verweis auf index.php statt /index.php)
```
- no errors in php_error.log
- my index.php isn't a symlink
- and I try to call the list with a commented out _.htaccess

**Comment by thomaszbz** [comment]
Then this is related to #147 (or maybe #129).

With #147, Ajax URLs are now generated by TYPO3 core (https://github.com/mblaschke/TYPO3-metaseo/issues/147#issuecomment-134793094). We don't deal with Ajax URL's any more in metaseo.

I think the Ajax URL (which was generated by TYPO3) looks right for a 6.2.19 instance. This URL then used by metaseo's javascript code in your browser to send Ajax requests. Still, the URL looks fine. However, your TYPO3 instance does not know how to deal with the request's URL.

Basically that means that you need to make `.htaccess` compliant with TYPO3's URL generation. Have a look at #147, I posted a few comments and links about TYPO3's deprecation concepts.

**Comment by thomaszbz** [comment]
Maybe you have the deprecation concept of TYPO3 CMS 7.6 enabled via `.htaccess` or vhost configuration? The relevant line is https://github.com/TYPO3/TYPO3.CMS/blob/TYPO3_7-6/typo3/.htaccess#L26

If the deprecation concept of TYPO3 7.6 is enabled for a TYPO3 6.2 instance, you might run into trouble.

**Comment by the-robman** [comment]
The RewriteRule not exists in my .htaccess. But I add the line and change the Rewrite target, to see what happend with the ajax call. Unfortunately comes the same 500 Error. I will ask the hoster about the deprecated.php.
Checklist:
- no RewriteRule for ajax
- no .htaccess in /typo3/ (I make an Symlink-Update to TYPO3 6.2.19 before)

Do you read this about my two Domain-Entries? If the ajax request deliver less than two results, is everything ok (status 200), I can see the entries in the url-list.
But if the request is greater than two json results, comes the 500 status code.
![two-ajaxcall-results-ok](https://cloud.githubusercontent.com/assets/16691428/13729959/6c9f6748-e941-11e5-8030-f26b6ba40c59.jpg)

**Comment by thomaszbz** [comment]
I understand. The screenshot is from the hosted system?

As long as _one_ request for ajax.php has a response with http status 200, then the deprecation concept of TYPO3 7.5 seems not to be configured (which is fine for 6.2). Otherwise you'd get something else (!= 200).

I'll try to reproduce your scenario with multiple (>2) subdomains. I'm not aware if we test that particular scenario frequently.

**Comment by thomaszbz** [comment]
It seems as if the Exception is not one of metaseo's `AjaxExceptions`. Instead, I believe it is thrown somewhere in the core. metaseo puts out the exception's message of alien exceptions in the response, together with http status 500: https://github.com/mblaschke/TYPO3-metaseo/blob/develop/Classes/Controller/AbstractAjaxController.php#L171

@the-robman What happens if you comment out exactly this line and clear caches? Is the response still covered up by the hoster? Maybe you even can provide a stacktrace?

If you could get some debug-output or the original output from the 500-response (which got covered-up by the hoster), that would help a lot. Maybe the hoster writes it back into a separate custom log file (before the cover-up)? Have you tried it in TYPO3's debug mode?

**Comment by the-robman** [comment]
Yes, the screenshot was from the hosted page. I comment out the line 171 and test the response in TYPO3 Configuration presets: Production and Development (all caches cleared). But nothing traced in deprecation_*.log or php_error.log. The error 500 page shows the same and I cant find any other (apache) logfiles on the hosted server. I create a second ticket at jweiland.net, with the question of error 500 logs. I am looking forward what they say.

**Comment by the-robman** [comment]
I call one guy from the jweiland support, and he look together with me (he reload the log and I produce the 500 error) into the apache logfiles. Result: nothing. No error, warning, infos... just nothing. Do you have more ideas from where the wind blows?

**Comment by thomaszbz** [comment]
You somehow need to get rid of the hoster's response to get the original response.

Please ask your hoster if they can switch off that cover-up response temporarily. Or if they can at least tell in which scenario the hoster's message appears.

Plan B is #229.

**Comment by the-robman** [comment]
I talk yesterday with the support and he send me this line:
```
www.xxx.de [Mon Mar 14 14:03:35 2016] [error] [client x.xx.x.xxx] Premature end of script headers: ajax.php, referer http://www.xxx.de/typo3/mod.php?M=MetaseoMetaseo_MetaseoSitemap&moduleToken=c19b418fb719979710cdca2fab8b1b49330a35ba&tx_metaseo_metaseometaseo_metaseositemap%5BpageId%5D=1&tx_metaseo_metaseometaseo_metaseositemap%5Baction%5D=sitemap&tx_metaseo_metaseometaseo_metaseositemap%5Bcontroller%5D=BackendSitemap
```
If I reduce the "pagingSize" in the "metaSeoConf" array to 30, can I see a result in the list. But the Web-Module produce the same error.

Another try was, to deaktivate "cooluri", clear table tx_metaseo_site, index some sites, ALL caches cleared, same error.

One more Info - warning, it's german :-)
Mail written to jweiland:
"ich habe bereits die "Stable" PHP Versionen 5.4, 5.5 und 5.6, jeweils in "Standard u. Extended", ausprobiert. Damit verbunden, die Vorgängerversion metaseo 1.0.8 komplett gelöscht und die neue Version 2.0.0 installiert. Die nötigen DB-Compares durchgeführt und auch die Tabelle tx_metaseo_sitemap geleert. Dabei alle Caches (Install Tool All Cache, FE-, Sys- u. allg. Cache) geleert, den typo3temp-Ordner gelöscht und neu, per Install Tool, initialisiert. Und immer wieder grüßt die Error 500 Seite..."

**Comment by thomaszbz** [comment]
Now, this is getting closer:

`Premature end of script headers: ajax.php`

https://wiki.knownhost.com/developmental/what-is-the-premature-end-of-script-headers-php-error

So there is a lot of possibilities.

Maybe check PHP script timeouts (fits together with higher probability to hit the bug when more rows are to be displayed)? Maybe try another PHP version (your hoster allows that!)? Maybe toggle with or without fastcgi? Maybe check server load? Maybe check memory consumption (what's your memory limit for php?)?

With this error message, it is probable that it needs to be fixed with server configuration. Maybe you should raise PHP script timeout and memory limit first.

**Comment by the-robman** [comment]
Oh my god! The solution is so simple... I change the PHP (5.2.28) version to "FastCGI" and it works! I asking now the support what are the differences between fastcgi and standard.

**Comment by thomaszbz** [comment]
Great. Happy TYPO3ing ;-)

**Comment by thomaszbz** [comment]
Seems to be a configuration problem. @rafu1987 describes how to mitigate it in #233

**Comment by thomaszbz** [comment]
Reopened #233 because reality check against major web servers indicates that we should limit the number of headers to make metaseo usable on major web servers.

**Comment by thomaszbz** [comment]
@the-robman would you like to test the patch in #233? With the patch applied, metaseo should also work with your previous setup.
